lib/spack/spack/hooks/check_dynamic_linking_elf.py

```diff
@@ -0,0 +1,172 @@
+    'ld-linux-x86-64',
+    'libc',
+    'libdl',
+    'libgcc_s',
+    'libm',
+    'libmemusage',
+    'libmvec',
+    'libnsl',
+    'libnss_compat',
+    'libnss_db',
+    'libnss_dns',
+    'libnss_files',
+    'libnss_hesiod',
+    'libpcprofile',
+    'libpthread',
+    'libresolv',
+    'librt',
+    'libSegFault',
+    'libthread_db',
+    'libutil',
+    # gcc
+    'libasan',
+    'libatomic',
+    'libcc1',
+    'libgcc_s',
+    'libgfortran',
+    'libgomp',
+    'libitm',
+    'liblsan',
+    'libquadmath',
+    'libssp',
+    'libstdc++',
+    'libtsan',
+    'libubsan',
+])
+
+
+def is_compatible(elf_a, elf_b):
+    return (elf_a.is_little_endian == elf_b.is_little_endian and
+            elf_a.is_64_bit == elf_b.is_64_bit and
+            elf_a.isa == elf_b.isa)
+
+
+def candidate_matches(current_elf, candidate_path):
+    try:
+        with open(candidate_path, 'rb') as g:
+            return is_compatible(current_elf, elf.parse_elf_file(g))
+    except (IOError, DeprecationWarning, struct.error, elf.ElfParsingError):
+        return False
+
+
+def read_elf(path):
+    try:
+        with open(path, 'rb') as f:
+            current_elf = elf.parse_elf_file(f)
+            if not current_elf.dynamic or len(current_elf.needed_offsets) == 0:
+                return None
+            needed = [
+                elf.read_c_string_from_file(f, off).decode("utf-8") for
+                off in current_elf.needed_offsets
+            ]
+            return current_elf, needed
+    except (IOError, DeprecationWarning, struct.error, elf.ElfParsingError):
+        return None
+
+
+def locate_needed(search_dirs, current_elf, needed, num_not_found):
+    """
+    Find libraries by <search dir>/<name>. At the end the needed
+    list is partitioned as [not found | found] and the number of
+    libraries that are not found is returned.
+    """
+    for dir in search_dirs:
+        i = 0
+        while i < num_not_found:
+            # Try to open by path
+            candidate = os.path.join(dir, needed[i])
+
+            if candidate_matches(current_elf, candidate):
+                num_not_found -= 1
+                needed[i], needed[num_not_found] = needed[num_not_found], needed[i]
+            else:
+                i += 1
+
+    return num_not_found
+
+
+def partition_needed(needed):
+    """
+    Partition the needed list in [soname | path/skipped]
+    """
+    num_sonames = len(needed)
+    i = 0
+    while i < num_sonames:
+        if '/' in needed[i] or needed[i].split(".so", 1)[0] in skip_list:
+            num_sonames -= 1
+            needed[i], needed[num_sonames] = needed[num_sonames], needed[i]
+        else:
+            i += 1
+    return num_sonames
+
+
+def get_search_paths(spec):
+    # Use self + direct link deps as search directories.
+    search_dirs = []
+    for s in [spec] + spec.dependencies(deptype='link'):
+        if s.external:
+            continue
+        try:
+            dirs = spec[s.name].libs.directories
```

Comment:
I think this is causing a CI failure:

```
spack.build_environment.ChildError: AttributeError: Query of package 'mpileaks' for 'libs' failed
```

(at https://github.com/spack/spack/runs/4598014455?check_suite_focus=true): 

* this is now calling `.libs` for every installed package
* the default implementation fails if no libs are found
* the mock packages don't have libraries

So one solution for example would be to define `.libs` for the mock `mpileaks` package.

Reply:
I think this hook should not run in tests

Reply:
Generally the build tests and CI jobs are sufficient to exercise this so I mostly agree. I think at least one unit test should run the hook though (so e.g. if something about the API of `.libs` changes, it is easy to see that the hook also needs to be updated).